### PR TITLE
Fix recovery stage transition with sync_id

### LINKED ISSUE #57708: Shard cannot be relocated after setting node exclusion.
We have met a shard relocation issue after setting node exclusion. In our case, original cluster is 6.8.2, we try to add the same amount of new 7.5.1 nodes and exclude the 6.8.2 nodes to upgrade cluster.

However, after adding 7.5.1 nodes, and set exclude 6.8.2 nodes in cluster setting, one of the single empty .kibana index shard cannot be relocated success, we have met this issue in several times.

Here is the node list after adding new nodes, we could see 4 6.8.2 nodes and 4 7.5.1 nodes:
```
[c_log@VM_1_14_centos ~/repository]$ curl "localhost:9200/_cat/nodes?h=version,name,node.role&s=version"
6.8.2 1590650188002472432 dmi
6.8.2 1590650188002472632 dmi
6.8.2 1590650188002472732 dmi
6.8.2 1590650188002472532 dmi
7.5.1 1590650759002483032 dmi
7.5.1 1590650759002483132 dmi
7.5.1 1590650759002482832 dmi
7.5.1 1590650759002482932 dmi
```

And we set this cluster setting to exclude data from 6.8.2:
```
"transient" : {
    "cluster" : {
      "routing" : {
        "allocation" : {
          "node_concurrent_recoveries" : "10",
          "exclude" : {
            "_name" : "1590650188002472632,1590650188002472732,1590650188002472432,1590650188002472532"
          }
        }
      }
    }
```

The cluster is empty and only contains kibana index. We could see the single internal .kibana_1 system index and it contains nothing docs:

```
[c_log@VM_1_14_centos ~]$ curl localhost:9200/_cat/indices?v
health status index uuid pri rep docs.count docs.deleted store.size pri.store.size
green open .kibana_1 5nRyca57QeaIN4O_SerQ7g 1 1 0 0 522b 261b
```
Finally, the shard 0 replica cannot be relocated to the new node:

```
[c_log@VM_1_14_centos ~]$ curl localhost:9200/_cat/shards?v
index shard prirep state docs store ip node
.kibana_1 0 p STARTED 0 261b 10.0.0.82 1590650759002483132 (relocated success)
.kibana_1 0 r STARTED 0 261b 10.0.0.148 1590650188002472732 (fail shard, it should be relocated)
```

On the master and target node, we could see this exception, no exception on source node:

```
 [2020-05-28T15:26:59,295][WARN ][o.e.i.c.IndicesClusterStateService] [1590650759002483032] [.kibana_1][0] marking and sending shard failed due to [failed recovery]
org.elasticsearch.indices.recovery.RecoveryFailedException: [.kibana_1][0]: Recovery failed from {1590650759002483132}{o5bJB_gPT6WiEDdt0l-v0Q}{AaaWa5nTQOuOwacnaE5xpA}{10.0.0.82}{10.0.0.82:20839}{di}{temperature=hot, rack=cvm_1_100003, set=100003, region=1, ip=9.10.49.143} into {1590650759002483032}{l42RGM6tSz-3-Dquma5OzQ}{ZZSxRSWXQjOFlETue5UHxQ}{10.0.0.205}{10.0.0.205:29559}{di}{rack=cvm_1_100003, set=100003, ip=9.10.48.33, temperature=hot, region=1}
        at org.elasticsearch.indices.recovery.PeerRecoveryTargetService.lambda$doRecovery$2(PeerRecoveryTargetService.java:247) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$1.handleException(PeerRecoveryTargetService.java:292) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.transport.PlainTransportFuture.handleException(PlainTransportFuture.java:97) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1120) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.transport.InboundHandler.lambda$handleException$2(InboundHandler.java:259) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703) [elasticsearch-7.5.1.jar:7.5.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_181]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_181]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_181]
Caused by: org.elasticsearch.transport.RemoteTransportException: [1590650759002483132][10.0.0.82:20839][internal:index/shard/recovery/start_recovery]
Caused by: java.lang.IllegalStateException: can't move recovery to stage [FINALIZE]. current stage: [INDEX] (expected [TRANSLOG])
        at org.elasticsearch.indices.recovery.RecoveryState.validateAndSetStage(RecoveryState.java:175) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.indices.recovery.RecoveryState.setStage(RecoveryState.java:206) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.index.shard.IndexShard.finalizeRecovery(IndexShard.java:1718) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.indices.recovery.RecoveryTarget.lambda$finalizeRecovery$1(RecoveryTarget.java:313) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:285) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.indices.recovery.RecoveryTarget.finalizeRecovery(RecoveryTarget.java:294) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$FinalizeRecoveryRequestHandler.messageReceived(PeerRecoveryTargetService.java:395) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.indices.recovery.PeerRecoveryTargetService$FinalizeRecoveryRequestHandler.messageReceived(PeerRecoveryTargetService.java:389) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:63) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.transport.InboundHandler$RequestHandler.doRun(InboundHandler.java:280) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:773) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-7.5.1.jar:7.5.1]
        ... 3 more
```

The cluster is in green status after relocating failed, just the shard cannot be relocated and remain on the excluding node. This issue could not be easily re-produced. 

The key log message is `can't move recovery to stage [FINALIZE]. current stage: [INDEX] (expected [TRANSLOG])`, it seems has any recovering process gap between 6.8 and 7.5.

### PR DESCRIPTION
**Pull request by dnhatn**
If the recovery source is on an old node (before 7.2), then the recovery target won't have the safe commit after phase1 because the recovery source does not send the [global checkpoint](https://github.com/elastic/elasticsearch/pull/40823) in the clean_files step. And if the recovery fails and retries, then the recovery stage won't transition properly. If a sync_id is used in peer recovery, then the clean_files step won't be executed to move the stage to TRANSLOG.

This issue was addressed in #57187, but not forward-ported to 8.0. I think we should do it as this issue can occur in 8.0. (requires a full cluster restart to 8.0 after a peer recovery on 7.1 fails after it has completed phase 1).

Closes #57708

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/gateway/ReplicaShardAllocatorSyncIdIT.java [hunk 1]
```diff
@@ -243,6 +245,53 @@ public void testFullClusterRestartPerformNoopRecovery() throws Exception {
         assertNoOpRecoveries(indexName);
     }
 
+    /**
+     * If the recovery source is on an old node (before 7.2), then the recovery target won't have the safe commit after phase1
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/gateway/ReplicaShardAllocatorSyncIdIT.java:249]: Can we drop this test in v9? If so, can we link `Version#V_7_2_0` in the comment to remind us?

**Review by DaveCTurner — COMMENTED** [review]
Is this assertion valid? If so, can we add it?

```diff
diff --git a/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java b/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
index 33139912920..8000bbabc53 100644
--- a/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
+++ b/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java
@@ -542,6 +542,8 @@ public class RecoverySourceHandler {
                         phase1ExistingFileSizes, existingTotalSize, took));
                 }, listener::onFailure);
             } else {
+                assert shard.indexSettings().getIndexVersionCreated().before(Version.V_7_2_0) ||
+                    request.startingSeqNo() != SequenceNumbers.UNASSIGNED_SEQ_NO;
                 logger.trace("skipping [phase1] since source and target have identical sync id [{}]", recoverySourceMetadata.getSyncId());

                 // but we must still create a retention lease
```

LGTM apart from that and one other request.

**Comment by howardhuanghua** [comment]
Hi @dnhatn , based on this comment:
> If the recovery source is on an old node (before 7.2), then the recovery target won't have the safe commit after phase1 because the recovery source does not send the [global checkpoint](https://github.com/elastic/elasticsearch/pull/40823) in the clean_files step. And if the recovery fails and retries, then the recovery stage won't transition properly. If a sync_id is used in peer recovery, then the clean_files step won't be executed to move the stage to TRANSLOG.

That means in recovering shard from pre-7.2 version to 7.2+ version, only if the recovery fails at the first time and retry again, then it cannot move to TRANSLOG stage? Is that possible it cannot move to TRANSLOG stage in a normal recovery process without any fail at the first time? Since in our case, 6.8 add 7.5 nodes to recovery #57708, we didn't see fails except the wrong stage exception. Also why after performing `_reroute?retry_failed` could relocate the stunk shard success? Thanks.

**Comment by howardhuanghua** [comment]
By the way, in #57708, both replica and primary are on 7.5.1 node. The flow is:
1. Exclude 6.8.2 version nodes.
2. `.kibana` primary shard relocate from 6.8.2 node to 7.5.1 node.
3. `.kibana` replica shard peer recovery from primary (both on 7.5.1), relocation blocked.

**Comment by dnhatn** [comment]
Thanks @howardhuanghua. I am trying to reproduce your situation.

**Comment by howardhuanghua** [comment]
@dnhatn Thanks, there is a point that in our situation, we could see the blocked index is empty, no docs contained.

**Comment by dnhatn** [comment]
Hi @howardhuanghua,

@hubbleview was kind to help me to reproduce the scenario that you provided. It matches what we outline in the PR. When migrating to new nodes, we first synced flush an index, then we exclude the old nodes in the allocation filter. ES will relocate both the primary and replica to the new nodes at the same time. If phase2 of the recovery of the replica starts after the relocation of the primary completes, then it will hit an IllegalIndexShardStateException. This exception is retry-able, hence we log it at the [trace](https://github.com/elastic/elasticsearch/blob/2f9b9d0ff5f1c246b38f65c27e60132f7d8b73fb/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L192) level and [retry](https://github.com/elastic/elasticsearch/blob/2f9b9d0ff5f1c246b38f65c27e60132f7d8b73fb/server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java#L219) another recovery. At this point, the primary is on the new node, and the replica has an index commit with a sync_id, but that commit is not safe (because we do not have the global checkpoint in the clean_files step). The first retry recovery will fail due to the improper transition, but the subsequent recovery will succeed as we should have the global checkpoint the [finalize_step](https://github.com/elastic/elasticsearch/blob/2f9b9d0ff5f1c246b38f65c27e60132f7d8b73fb/server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java#L293).

Thank you for reporting the issue.

File: server/src/internalClusterTest/java/org/elasticsearch/gateway/ReplicaShardAllocatorSyncIdIT.java (hunk 1 above)
Comment (dnhatn) [reply server/src/internalClusterTest/java/org/elasticsearch/gateway/ReplicaShardAllocatorSyncIdIT.java]: ++. I pushed https://github.com/elastic/elasticsearch/pull/57754/commits/b8d63fb4a00ab5958df4adbd710994d3fc92bc27.

**Comment by dnhatn** [comment]
@DaveCTurner Thanks for reviewing. The assertion is great, but we will trip it if we hit a simulated I/O exception while we are [recovering locally](https://github.com/elastic/elasticsearch/blob/e9d13ec6291c662034b14e54517de312762b4c01/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1437).

**Comment by howardhuanghua** [comment]
Hi @dnhatn, thanks for the explanation, now I understood the issue. Just want to confirm the relocation scenarios, 
> ES will relocate both the primary and replica to the new nodes at the same time.
1. Set allocation filter, both primary and replica would be relocated to new nodes at the same time as you described.
2. Use shard move to move a replica to another node, this replica would be peer recovered from its primary or the original replica?
3. If replica shard failed itself and no local data could be used, it always peer recovery from its primary?

**Comment by dnhatn** [comment]
Hi @howardhuanghua,

> 1. Set allocation filter, both primary and replica would be relocated to new nodes at the same time as you described.

Yes, that's correct. 

> 2. Use shard move to move a replica to another node, this replica would be peer recovered from its primary or the original replica?

A replica always recovers from its primary.

> 3. If replica shard failed itself and no local data could be used, it always peer recovery from its primary?

A replica always recovers from its primary, and the recovery tries to reuse the existing data when possible.

**Comment by DaveCTurner** [comment]
> The assertion is great, but we will trip it if we hit a simulated I/O exception while we are recovering locally.

Is that ok? If we couldn't recover the shard locally does it make sense to proceed with the rest of the recovery like that?

**Comment by dnhatn** [comment]
> > The assertion is great, but we will trip it if we hit a simulated I/O exception while we are recovering locally.

> Is that ok? If we couldn't recover the shard locally does it make sense to proceed with the rest of the recovery like that?

When the local translog is corrupted, we won't be able to recover locally up to the global checkpoint. In this case, we will try to reuse the existing index commit, which can have a sync_id.

**Review by DaveCTurner — APPROVED** [review]
Ok, since the whole sync marker thing is going away soon we don't need to dwell on this further so this LGTM.

**Comment by dnhatn** [comment]
Thanks David!

**Pull request merged**